drivers/pwm/pwm_bflb_2.c

```diff
@@ -0,0 +1,199 @@
+#include <glb_reg.h>
+#include <common_defines.h>
+#include <zephyr/dt-bindings/clock/bflb_clock_common.h>
+#include <bouffalolab/common/pwm_v2_reg.h>
+
+#define PWM_WAIT_TIMEOUT_MS	100
+#define PWM_CH_OFFSET_MUL	4
+#define PWM_CH_POLARITY_MUL	2
+#define CHANNELS		4
+
+struct pwm_bflb_config {
+	uintptr_t base;
+	const struct pinctrl_dev_config *pcfg;
+};
+
+struct pwm_bflb_data {
+	uint32_t period_cycles;
+};
+
+static int pwm_bflb_get_cycles_per_sec(const struct device *dev, uint32_t ch, uint64_t *cycles)
+{
+	const struct device *clock_ctrl =  DEVICE_DT_GET_ANY(bflb_clock_controller);
+	uint32_t clk;
+	int ret;
+
+	ret = clock_control_get_rate(clock_ctrl, (void *)BFLB_CLKID_CLK_BCLK, &clk);
+
+	*cycles = clk / 2;
+
+	LOG_DBG("cycles: %d", clk);
+
+	return ret;
+}
+
+static int pwm_bflb_detrigger(const struct device *dev)
+{
+	const struct pwm_bflb_config *cfg = dev->config;
+	volatile uint32_t tmp;
+
+	tmp = sys_read32(cfg->base + PWM_MC0_CONFIG0_OFFSET);
+	tmp |= PWM_STOP_EN;
+	sys_write32(tmp, cfg->base + PWM_MC0_CONFIG0_OFFSET);
+
+	return 0;
+}
+
+
+static int pwm_bflb_trigger(const struct device *dev)
+{
+	const struct pwm_bflb_config *cfg = dev->config;
+	volatile uint32_t tmp;
+
+	tmp = sys_read32(cfg->base + PWM_MC0_CONFIG0_OFFSET);
+	tmp &= ~PWM_STOP_EN;
+	sys_write32(tmp, cfg->base + PWM_MC0_CONFIG0_OFFSET);
+
+	return 0;
+}
+
+static int pwm_bflb_set_cycles(const struct device *dev, uint32_t ch, uint32_t period_cycles,
+			   uint32_t pulse_cycles, pwm_flags_t flags)
+{
+	const struct pwm_bflb_config *cfg = dev->config;
+	struct pwm_bflb_data *data = dev->data;
+	k_timepoint_t end_timeout =
+		sys_timepoint_calc(K_MSEC(PWM_WAIT_TIMEOUT_MS));
+	volatile uint32_t tmp;
+	uint32_t pulse;
+	uint16_t divider, period;
+
+	if (ch >= CHANNELS) {
+		return -EINVAL;
+	}
+
+	divider = period_cycles / UINT16_MAX + 1;
+
+	period = period_cycles / divider;
+
+	if (period_cycles != data->period_cycles) {
+		LOG_INF("Changing global period!");
+		pwm_bflb_detrigger(dev);
+
+		do {
+			tmp = sys_read32(cfg->base + PWM_MC0_CONFIG0_OFFSET);
+		} while (!(tmp & PWM_STS_STOP) && !sys_timepoint_expired(end_timeout));
+		if (sys_timepoint_expired(end_timeout)) {
+			return -ETIMEDOUT;
+		}
+
+		tmp = sys_read32(cfg->base + PWM_MC0_CONFIG0_OFFSET);
+		tmp &= ~PWM_CLK_DIV_MASK;
+		tmp |= (uint32_t)divider << PWM_CLK_DIV_SHIFT;
+		sys_write32(tmp, cfg->base + PWM_MC0_CONFIG0_OFFSET);
+
+		tmp = sys_read32(cfg->base + PWM_MC0_PERIOD_OFFSET);
+		tmp &= ~PWM_PERIOD_MASK;
+		tmp |= (uint32_t)period << PWM_PERIOD_SHIFT;
+		sys_write32(tmp, cfg->base + PWM_MC0_PERIOD_OFFSET);
+	}
+
+	pulse = pulse_cycles / divider;
+
+	LOG_DBG("divider: %d period: %d pulse: %d", divider, period, pulse);
+
+	tmp = pulse << PWM_CH0_THREH_SHIFT;
+	sys_write32(tmp, cfg->base + PWM_MC0_CH0_THRE_OFFSET + ch * PWM_CH_OFFSET_MUL);
+
+	tmp = sys_read32(cfg->base + PWM_MC0_CONFIG1_OFFSET);
+	tmp |= PWM_CH0_PEN << (PWM_CH_OFFSET_MUL * ch);
+	tmp |= PWM_CH0_NEN << (PWM_CH_OFFSET_MUL * ch);
```

Comment:
I got very slightly confused by `PWM_CH_OFFSET_MUL` used for both the address offset and bit offset, but it's all correct value-wise. Could be a different name to avoid potential footguns.

```c
#define PWM_CH_SHIFT_MUL	4
```

```suggestion
	tmp |= PWM_CH0_PEN << (PWM_CH_SHIFT_MUL * ch);
	tmp |= PWM_CH0_NEN << (PWM_CH_SHIFT_MUL * ch);
```

Still makes sense to me as it is though.